tests/runtime-tests/src/services/docker.rs

```diff
@@ -0,0 +1,164 @@
+use super::Service;
+use anyhow::{bail, Context as _};
+use std::{
+    cell::OnceCell,
+    collections::HashMap,
+    path::Path,
+    process::{Command, Stdio},
+};
+
+/// A docker container as a service
+pub struct DockerService {
+    image_name: String,
+    container: Container,
+    // We declare lock after container so that the lock is dropped after the container is
+    _lock: fslock::LockFile,
+    ports: OnceCell<HashMap<u16, u16>>,
+}
+
+impl DockerService {
+    /// Start a docker container as a service
+    pub fn start(name: &str, service_definitions_path: &Path) -> anyhow::Result<Self> {
+        let docker_file_path = service_definitions_path.join(format!("{name}.Dockerfile"));
+        let image_name = format!("spin/runtime-tests/services/{name}");
+        let mut lock =
+            fslock::LockFile::open(&service_definitions_path.join(format!("{name}.lock")))
+                .context("failed to open service file lock")?;
+        lock.lock().context("failed to obtain service file lock")?;
+
+        stop_containers(&get_running_containers(&image_name)?)?;
+        build_image(&docker_file_path, &image_name)?;
+        let container = run_container(&image_name)?;
+
+        Ok(Self {
+            image_name,
+            container,
+            _lock: lock,
+            ports: OnceCell::new(),
+        })
+    }
+}
+
+struct Container {
+    id: String,
+}
+
+impl Container {
+    fn get_ports(&self) -> anyhow::Result<HashMap<u16, u16>> {
+        let output = Command::new("docker")
+            .arg("port")
+            .arg(&self.id)
+            .output()
+            .context("docker failed to fetch ports")?;
+        if !output.status.success() {
+            bail!("failed to run fetch ports for docker container");
+        }
+        let output = String::from_utf8(output.stdout)?;
+        output
+            .lines()
+            .map(|s| {
+                // 3306/tcp -> 0.0.0.0:32770
+                let s = s.trim();
+                let (guest, host) = s
+                    .split_once(" -> ")
+                    .context("failed to parse port mapping")?;
+                let (guest_port, _) = guest.split_once('/').context("TODO")?;
+                let host_port = host.rsplit(':').next().context("TODO")?;
+                Ok((guest_port.parse()?, host_port.parse()?))
+            })
+            .collect()
+    }
+}
+
+impl Drop for Container {
+    fn drop(&mut self) {
+        let _ = stop_containers(&[std::mem::take(&mut self.id)]);
+    }
+}
+
+impl Service for DockerService {
```

Comment:
Fwiw you may want to look at https://github.com/testcontainers/testcontainers-rs - at least in Go they've got some useful helpers for things like "waiting for a service to actually be up"

Reply:
I implemented health checking now too which was only a few lines of code, so it seems like we're fine without a library for now.